### ISSUE TITLE
Data Source "vault_generic_secret" fails to parse keys with dots in the name

### ISSUE THREAD

**Issue by legal90** [terraform-providers/terraform-provider-vault#360]
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

### Versions
Terraform v0.11.13
+ provider.vault v1.6.0

Vault Server v1.0.1
### Affected Resource(s)
Please list the resources as a list, for example:
- `data.vault_generic_secret`

### Terraform Configuration Files
```hcl
# ...

data "vault_generic_secret" "aws" {
  path = "secret/my_dir/aws_credentials"
}

locals {
  secret_key = "${data.vault_generic_secret.aws.data["certmanager.aws_secret_access_key"]}"
}
```
### Vault Secret
```hcl
$ vault kv get secret/my_dir/aws_credentials

====== Metadata ======
Key              Value
---              -----
created_time     2019-03-17T08:35:30.579496052Z
deletion_time    n/a
destroyed        false
version          10

================== Data ==================
Key                                  Value
---                                  -----
certmanager.aws_secret_access_key    mySecretAccessKey
foo                                  bar
```

The Vault backend is **KV2** with versioning enabled.

### Expected Behavior
The data key `certmanager.aws_secret_access_key` is fetched from the Vault key

### Actual Behavior
```
Error: Error refreshing state: 1 error(s) occurred:

* local.secret_key: local.secret_key: At column 37, line 1: map "data.vault_generic_secret.aws.data" does not have homogenous types. found TypeMap and then TypeString in:

${data.vault_generic_secret.aws.data["certmanager.aws_secret_access_key"]}

```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform plan`

### Important Factoids
In my example the Vault secret has 2 keys - one with a dot and another one without a dot in the key name. 
However, if there are only keys with dots, then the terraform plan fails with another error:
```
* local.secret_key: local.secret_key: key "certmanager.aws_secret_access_key" does not exist in map data.vault_generic_secret.aws.data in:

${data.vault_generic_secret.aws.data["certmanager.aws_secret_access_key"]}
```

It seems that the provider parses keys with dots as a map structure, instead of a string.

### References
- https://github.com/hashicorp/consul-template/issues/348 (I'm not sure that it is the same root cause, but the effect is similar)

My actual configuration and debug output contains a lot of sensitive information. I hope that details I provided above are enough to reproduce the issue on a simple configuration. But if it's not - please let me know, I will prepare a more complete reproduction scenario.

**Comment by legal90** [comment]
It seems that the issue is actually on the terraform side:
- https://github.com/hashicorp/terraform/issues/12546
- https://github.com/hashicorp/terraform/issues/19258

Is there any workaround for the Vault use case?

**Comment by estein9825** [comment]
Did you try a backslash?

**Comment by aquabubble** [comment]
> Did you try a backslash? Or a double-backslash if it requires the backslash to filter through to the vault resource.

The issue occurs if the secret contains any variables with a period in its keys even if the data being retrieved does not.

**Comment by legal90** [comment]
@estein9825 Yes. The escaping doesn't help because the error happens on the evaluation phase, when terraform reads the structure of the secret and tries to parse it to its internal map structure.

So, it really looks like a limitation of terraform :(

**Comment by lmickh** [comment]
Can confirm this is a Terraform type error.  I came across it awhile ago.  There should be even more issues related to it.  There is not a good solution except to avoid using periods in key names that you wish to access with Terraform.
